std/experimental/checkedint.d

```diff
@@ -0,0 +1,1626 @@
+        }
+        else static if (x == "-")
+        {
+            if (v != defaultValue!T) --v;
+        }
+    }
+    auto hookOpBinary(string x, L, R)(L lhs, R rhs)
+    {
+        alias Result = typeof(lhs + rhs);
+        return lhs != defaultValue!L
+            ? mixin("lhs"~x~"rhs")
+            : defaultValue!Result;
+    }
+    auto hookOpBinaryRight(string x, L, R)(L lhs, R rhs)
+    {
+        alias Result = typeof(lhs + rhs);
+        return rhs != defaultValue!R
+            ? mixin("lhs"~op~"rhs")
+            : defaultValue!Result;
+    }
+    void hookOpOpAssign(string x, L, R)(ref L lhs, R rhs)
+    {
+        if (lhs != defaultValue!L) mixin("lhs"~x~"=rhs;");
+    }
+}
+
+///
+unittest
+{
+    auto x1 = Checked!(int, WithNaN)();
+    assert(x1.payload == int.min);
+    assert(x1 != x1);
+    assert(!(x1 < x1));
+    assert(!(x1 > x1));
+    assert(!(x1 == x1));
+    ++x1;
+    assert(x1.payload == int.min);
+    --x1;
+    assert(x1.payload == int.min);
+    x1 = 42;
+    assert(x1 == x1);
+    assert(x1 <= x1);
+    assert(x1 >= x1);
+    static assert(x1.min == int.min + 1);
+    x1 += long(int.max);
+}
+
+unittest
+{
+    alias Smart(T) = Checked!(Checked!(T, ProperCompare), WithNaN);
+    Smart!int x1;
+    assert(x1 != x1);
+    x1 = -1;
+    assert(x1 < 1u);
+    auto x2 = Smart!int(42);
+}
+
+/*
+Yields `true` if `T1` is "value convertible" (using terminology from C) to
+`T2`, where the two are integral types. That is, all of values in `T1` are
+also in `T2`. For example `int` is value convertible to `long` but not to
+`uint` or `ulong`.
+*/
+/*
+private enum valueConvertible(T1, T2) = isIntegral!T1 && isIntegral!T2 &&
+    is(T1 : T2) && (
+        isUnsigned!T1 == isUnsigned!T2 || // same signedness
+        !isUnsigned!T2 && T2.sizeof > T1.sizeof // safely convertible
+    );
+*/
+template valueConvertible(T1, T2)
+{
+    static if (!isIntegral!T1 || !isIntegral!T2)
+    {
+        enum bool valueConvertible = false;
+    }
+    else
+    {
+        enum bool valueConvertible = is(T1 : T2) && (
+            isUnsigned!T1 == isUnsigned!T2 || // same signedness
+            !isUnsigned!T2 && T2.sizeof > T1.sizeof // safely convertible
+        );
+    }
+}
+
+/**
+
+Defines binary operations with overflow checking for any two integral types.
+The result type obeys the language rules (even when they may be
+counterintuitive), and `overflow` is set if an overflow occurs (including
+inadvertent change of signedness, e.g. `-1` is converted to `uint`).
+Conceptually the behavior is:
+
+$(OL $(LI Perform the operation in infinite precision)
+$(LI If the infinite-precision result fits in the result type, return it and
+do not touch `overflow`)
+$(LI Otherwise, set `overflow` to `true` and return an unspecified value)
+)
+
+The implementation exploits properties of types and operations to minimize
+additional work.
+
+*/
+typeof(L() + R()) opChecked(string x, L, R)(const L lhs, const R rhs,
+    ref bool overflow)
```

Comment:
`overflow` should be changed to something more generic, as (for example) `0 / 0` does not overflow, and yet should signal an error.

Suggestions: `error`, `fail`, `NaN`, `invalid`, `unrepresentable`